// @typegoose/auto-increment/lib/autoIncrement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoIncrementID = exports.AutoIncrementIDSkipSymbol = exports.AutoIncrementSimple = exports.isNullOrUndefined = void 0;
const tslib_1 = require("tslib");
const mongoose = require("mongoose");
const logSettings_1 = require("./logSettings");
const DEFAULT_INCREMENT = 1;
/**
 * Because since node 4.0.0 the internal util.is* functions got deprecated
 * @param val Any value to test if null or undefined
 */
function isNullOrUndefined(val) {
    return val === null || val === undefined;
}
exports.isNullOrUndefined = isNullOrUndefined;
/**
 * The Plugin - Simple version
 * Increments an value each time it is saved
 * @param schema The Schema
 * @param options The Options
 */
function AutoIncrementSimple(schema, options) {
    // convert normal object into an array
    const fields = Array.isArray(options) ? options : [options];
    logSettings_1.logger.info('Initilaize AutoIncrement for an schema with %d fields to increment', fields.length);
    if (fields.length <= 0) {
        throw new Error('Options with at least one field are required!');
    }
    // check if all fields are valid
    for (const field of fields) {
        if (typeof field.field != 'string') {
            throw new Error(`Got Field object, but did not contain a "field" key with "string" value, got: ${JSON.stringify(field)}`);
        }
        const schemaField = schema.path(field.field);
        // check if the field is even existing
        if (isNullOrUndefined(schemaField)) {
            throw new Error(`Field "${field.field}" does not exists on the Schema!`);
        }
        // check if the field is an number
        if (schemaField.instance !== 'Number' && schemaField.instance !== 'BigInt') {
            throw new Error(`Field "${field.field}" is not a Number or BigInt!`);
        }
        if (isNullOrUndefined(field.incrementBy)) {
            logSettings_1.logger.info('Field "%s" does not have an incrementBy defined, defaulting to %d', field.field, DEFAULT_INCREMENT);
            if (schemaField.instance === 'BigInt') {
                field.incrementBy = BigInt(DEFAULT_INCREMENT);
            }
            else {
                field.incrementBy = DEFAULT_INCREMENT;
            }
        }
    }
    // to have an name to the function if debugging
    schema.pre('save', function AutoIncrementPreSaveSimple() {
        if (!this.isNew) {
            logSettings_1.logger.info('Starting to increment "%s"', this.constructor.modelName);
            for (const field of fields) {
                logSettings_1.logger.info('Incrementing "%s" by %d', field.field, field.incrementBy);
                // "this as any" is required because mongoose sets a document to default to "unknown" if not explicitly set for hooks
                // but in this case it should work just fine
                this[field.field] += field.incrementBy;
            }
        }
    });
}
exports.AutoIncrementSimple = AutoIncrementSimple;
/** The Schema used for the trackers */
const IDSchema = new mongoose.Schema({
    field: String,
    modelName: String,
    count: BigInt,
}, { versionKey: false });
IDSchema.index({ field: 1, modelName: 1 }, { unique: true });
exports.AutoIncrementIDSkipSymbol = Symbol('AutoIncrementIDSkip');
/**
 * The Plugin - ID
 * Increments an counter in an tracking collection
 * @param schema The Schema
 * @param options The Options
 */
function AutoIncrementID(schema, options) {
    /** The Options with default options applied */
    const opt = {
        field: '_id',
        incrementBy: DEFAULT_INCREMENT,
        trackerCollection: 'identitycounters',
        trackerModelName: 'identitycounter',
        startAt: 0,
        overwriteModelName: '',
        ...options,
    };
    // check if the field is an number
    const schemaField = schema.path(opt.field);
    if (schemaField.instance !== 'Number' && schemaField.instance !== 'BigInt') {
        throw new Error(`Field "${opt.field}" is not a Number or BigInt!`);
    }
    let model;
    logSettings_1.logger.info('AutoIncrementID called with options %O', opt);
    schema.pre('save', async function AutoIncrementPreSaveID() {
        logSettings_1.logger.info('AutoIncrementID PreSave');
        opt.incrementBy = BigInt(opt.incrementBy);
        opt.startAt = BigInt(opt.startAt);
        const originalModelName = this.constructor.modelName;
        let modelName;
        if (typeof opt.overwriteModelName === 'function') {
            modelName = opt.overwriteModelName(originalModelName, this.constructor);
            if (!modelName || typeof modelName !== 'string') {
                throw new Error('"overwriteModelname" is a function, but did return a falsy type or is not a string!');
            }
        }
        else {
            modelName = opt.overwriteModelName || originalModelName;
        }
        if (!model) {
            logSettings_1.logger.info('Creating idtracker model named "%s"', opt.trackerModelName);
            // needs to be done, otherwise "undefiend" error if the plugin is used in an sub-document
            const db = this.db ?? this.ownerDocument().db;
            model = db.model(opt.trackerModelName, IDSchema, opt.trackerCollection);
            // test if the counter document already exists
            const counter = await model
                .findOne({
                modelName: modelName,
                field: opt.field,
            })
                .lean()
                .exec();
            if (!counter) {
                await model.create({
                    modelName: modelName,
                    field: opt.field,
                    count: opt.startAt - opt.incrementBy,
                });
            }
        }
        if (!this.isNew) {
            logSettings_1.logger.info('Document is not new, not incrementing');
            return;
        }
        // @ts-expect-error mongoose now restrics indexes to "string"
        if (typeof this[exports.AutoIncrementIDSkipSymbol] === 'boolean' && exports.AutoIncrementIDSkipSymbol) {
            logSettings_1.logger.info('Symbol "AutoIncrementIDSkipSymbol" is set to "true", skipping');
            return;
        }
        const ownerDocDb = this.db ?? this.ownerDocument().db;
        if (typeof ownerDocDb[exports.AutoIncrementIDSkipSymbol] == 'boolean' && ownerDocDb[exports.AutoIncrementIDSkipSymbol]) {
            logSettings_1.logger.info('Symbol "AutoIncrementIDSkipSymbol" is set on the connection to "true", skipping');
            return;
        }
        const leandoc = await model
            .findOneAndUpdate({
            field: opt.field,
            modelName: modelName,
        }, {
            $inc: { count: opt.incrementBy },
        }, {
            new: true,
            fields: { count: 1, _id: 0 },
            upsert: true,
            setDefaultsOnInsert: true,
        })
            .lean()
            .exec();
        if (isNullOrUndefined(leandoc)) {
            throw new Error(`"findOneAndUpdate" incrementing count failed for "${modelName}" on field "${opt.field}"`);
        }
        logSettings_1.logger.info('Setting "%s" to "%d"', opt.field, leandoc.count);
        this[opt.field] = leandoc.count;
        return;
    });
}
exports.AutoIncrementID = AutoIncrementID;
tslib_1.__exportStar(require("./types"), exports);
